### DISCUSSION TITLE
DataGrid/DataGridDataGridCollectionView - Grouping performacne

### DISCUSSION THREAD

**Discussion by KrzysztofDusko** [AvaloniaUI/Avalonia#14475] [Q&A] [answered]
I have faced very bad performance when if grouping by column with many distinct values.
it turned out that IsDataInGroupOrder is efectivly always false.. wittch leads to O(ROWS_COUNT * GROUPS COUNT) complexity **even** when data is sorted by group column .. (each row i checked for all groups. not only actual and next)
I there anyone know why this decision was made ? 

its fixed to false in DataGridCollectionView contructor
https://github.com/AvaloniaUI/Avalonia/blob/c2798ce4adf4cb0f5a0fe0c89182429e9bc00502/src/Avalonia.Controls.DataGrid/Collections/DataGridCollectionView.cs#L201-L202

and ALL PrepareGroups() cals have IsDataInGroupOrder = false.. 
https://github.com/AvaloniaUI/Avalonia/blob/c2798ce4adf4cb0f5a0fe0c89182429e9bc00502/src/Avalonia.Controls.DataGrid/Collections/DataGridCollectionView.cs#L3119

and later on we have final O(ROWS_COUNT * GROUPS COUNT) loop.. 
https://github.com/AvaloniaUI/Avalonia/blob/c2798ce4adf4cb0f5a0fe0c89182429e9bc00502/src/Avalonia.Controls.DataGrid/Collections/DataGridGroupDescription.cs#L1063-L1080

**Comment by KrzysztofDusko** [comment]
only workaround I have is to modify _flags by reflection..

**Comment by timunie** [comment]
I believe it was not done by intention. The DataGrid was made long time ago when the team was way smaller and only a hobby project. So worth to file a PR to make it respect the value set if anyone wants to do it.

**Reply by KrzysztofDusko** [reply]
data grid code is pretty complex.. Will be accepted partial solution - addind methods to access _flags direct- maybe something like SetDataSorted(true/false) to change this flag temporay ? 
I asking becouse changing this  in full scope is very risky. and definitly code should be well tested witch is not so easy.. 
Of course reflection is "ok" for this as well

**Reply by KrzysztofDusko** [reply, marked as answer]
:( few lines below is another contructor..
![image](https://github.com/AvaloniaUI/Avalonia/assets/69449360/281713f1-0803-4159-8814-dade68d690c6)
